torchvision/models/swin_transformer.py

```diff
@@ -276,39 +303,89 @@ def forward(self, x: Tensor):
+        in_channels: int = 3,
+        norm_layer: Optional[Callable[..., nn.Module]] = None,
+    ):
+        super().__init__()
+        self.patch_size = patch_size
+        self.proj = nn.Conv2d(
+            in_channels, embed_dim, kernel_size=(patch_size[0], patch_size[1]), stride=(patch_size[0], patch_size[1])
+        )
+        self.norm: Optional[nn.Module]
+        if norm_layer is not None:
+            self.norm = norm_layer(embed_dim)
+        else:
+            self.norm = None
+
+    def forward(self, x: Tensor) -> Tensor:
+        """
+        Args:
+            x (Tensor): batch of images with layout [B, C, H, W]
+        Returns:
+            Tensor with layout [B, H/patch_size[0], W/patch_size[1], embed_dim]
+        """
+        _, _, H, W = x.shape
+        pad_r = (self.patch_size[1] - W % self.patch_size[1]) % self.patch_size[1]
+        pad_b = (self.patch_size[0] - H % self.patch_size[0]) % self.patch_size[0]
+        x = F.pad(x, (0, pad_r, 0, pad_b))
+
+        x = self.proj(x)
+        x = x.permute(0, 2, 3, 1)
+        if self.norm is not None:
+            x = self.norm(x)
+        return x
+
+
 class SwinTransformer(nn.Module):
     """
     Implements Swin Transformer from the `"Swin Transformer: Hierarchical Vision Transformer using
     Shifted Windows" <https://arxiv.org/pdf/2103.14030>`_ paper.
     Args:
-        patch_size (int): Patch size.
+        patch_size (List[int]): Patch size.
         embed_dim (int): Patch embedding dimension.
         depths (List(int)): Depth of each Swin Transformer layer.
         num_heads (List(int)): Number of attention heads in different layers.
-        window_size (int): Window size. Default: 7.
+        window_size (List[int]): Window size.
         mlp_ratio (float): Ratio of mlp hidden dim to embedding dim. Default: 4.0.
         dropout (float): Dropout rate. Default: 0.0.
         attention_dropout (float): Attention dropout rate. Default: 0.0.
         stochastic_depth_prob (float): Stochastic depth rate. Default: 0.0.
-        num_classes (int): Number of classes for classification head. Default: 1000.
+        num_classes (int, optional): Number of classes for classification head,
+            if None then the model have no head. Default: 1000.
         block (nn.Module, optional): SwinTransformer Block. Default: None.
         norm_layer (nn.Module, optional): Normalization layer. Default: None.
+        patch_embed (nn.Module, optional): Patch Embedding layer. Default: None.
     """
 
     def __init__(
         self,
-        patch_size: int,
+        patch_size: List[int],
         embed_dim: int,
         depths: List[int],
         num_heads: List[int],
-        window_size: int = 7,
+        window_size: List[int],
         mlp_ratio: float = 4.0,
         dropout: float = 0.0,
         attention_dropout: float = 0.0,
         stochastic_depth_prob: float = 0.0,
-        num_classes: int = 1000,
+        num_classes: Optional[int] = 1000,
```

Comment:
We make `num_classes` as optional, and it will not have any head if we set it None during instantiation.
This would be useful for use-cases that don't need head or need to switch head externally

Reply:
I don't think this API proposal should be part of this PR. It introduces something not supported by any other model. I would recommend examining this separately across the library.